spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/CatalogSharedInformerFactory.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2019-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.kubernetes.client.discovery;
+
+import io.kubernetes.client.informer.SharedInformerFactory;
+import io.kubernetes.client.openapi.models.V1Endpoints;
+import io.kubernetes.client.openapi.models.V1EndpointsList;
+import io.kubernetes.client.openapi.models.V1Service;
+import io.kubernetes.client.openapi.models.V1ServiceList;
+import io.kubernetes.client.spring.extended.controller.annotation.GroupVersionResource;
+import io.kubernetes.client.spring.extended.controller.annotation.KubernetesInformer;
+import io.kubernetes.client.spring.extended.controller.annotation.KubernetesInformers;
+
+@KubernetesInformers({
```

Comment:
this is not an inner class anymore, because it is used in both blocking and reactive configs. What is more interesting, is that if you look in `main` branch now (before this PR), `CatalogSharedInformerFactory` was declared but never used. We were having this:

```
@ConditionalOnReactiveDiscoveryEnabled
@AutoConfigureAfter(KubernetesDiscoveryClientAutoConfiguration.class)
public class KubernetesInformerReactiveDiscoveryClientAutoConfiguration {

      @Bean
	  @ConditionalOnMissingBean
	  public KubernetesInformerReactiveDiscoveryClient kubernetesReactiveDiscoveryClient(
			KubernetesNamespaceProvider kubernetesNamespaceProvider, SharedInformerFactory sharedInformerFactory ... {

       }

// --> This one over here : declared, but never actually used as a @Bean
      	class CatalogSharedInformerFactory extends SharedInformerFactory {

		          // TODO: optimization to ease memory pressure from continuous list&watch.

	     }

}
```

And in `KubernetesDiscoveryClientAutoConfiguration`:

```
public class KubernetesDiscoveryClientAutoConfiguration {
        @Configuration(proxyBeanMethods = false)
	    @ConditionalOnBlockingDiscoveryEnabled
	    public static class KubernetesInformerDiscoveryConfiguration {
               		@Bean
		            @ConditionalOnMissingBean
		            public CatalogSharedInformerFactory catalogSharedInformerFactory(ApiClient apiClient) {
			              return new CatalogSharedInformerFactory();
		            }
        }
}
```

Now imagine that someone does `spring.cloud.discovery.blocking.enabled=false`, so that `@ConditionalOnBlockingDiscoveryEnabled` is not resolved. `KubernetesInformerDiscoveryConfiguration` would not be picked-up, as such no `CatalogSharedInformerFactory` would be provided and `KubernetesInformerReactiveDiscoveryClientAutoConfiguration` would fail.

---

Because of that, I moved this class to a separate file and use it in both reactive and blocking implementations, with `@ConditionalOnMissingBean`. It is a lengthly explanation and may be not completely easy to follow, but we need (imho) such clean-ups in order to not be taken by surprise. At least now, like for all other changes that I made, I can take ownership and respond to issues/questions.